src/java.base/share/classes/jdk/internal/misc/MainMethodFinder.java

```diff
@@ -0,0 +1,164 @@
+                // no arguments or one string array argument.
+                if ("main".equals(method.getName()) &&
+                        !isPrivate(method) &&
+                        correctArgs(method) &&
+                        // Only statics in the declaring class
+                        (!isStatic(method) || declc == refc)
+                ) {
+                    mains.add(method);
+                }
+            }
+        }
+    }
+
+    /**
+     * Comparator for two methods.
+     * Priority order is;
+     * static < non-static,
+     * public < non-public,
+     * string arg < no arg and
+     * sub-class < super-class.
+     *
+     * @param a  first method
+     * @param b  second method
+     *
+     * @return -1, 0 or 1 to represent higher priority. equals priority or lesser priority.
+     */
+    private static int compareMethods(Method a, Method b) {
+        int aMods = a.getModifiers();
+        int bMods = b.getModifiers();
+        boolean aIsStatic = Modifier.isStatic(aMods);
+        boolean bIsStatic = Modifier.isStatic(bMods);
+
+        if (aIsStatic && !bIsStatic) {
+            return -1;
+        } else if (bIsStatic && !aIsStatic) {
+            return 1;
+        }
+
+        boolean aIsPublic = Modifier.isPublic(aMods);
+        boolean bIsPublic = Modifier.isPublic(bMods);
+
+        if (aIsPublic && !bIsPublic) {
+            return -1;
+        } else if (bIsPublic && !aIsPublic) {
+            return 1;
+        }
+
+        int aCount = a.getParameterCount();
+        int bCount = b.getParameterCount();
+
+        if (aCount > bCount) {
+            return -1;
+        } else if (bCount > aCount) {
+            return 1;
+        }
+
+        Class<?> aClass = a.getDeclaringClass();
+        Class<?> bClass = b.getDeclaringClass();
+
+        if (bClass.isAssignableFrom(aClass)) {
+            return -1;
+        } else if (bClass.isAssignableFrom(aClass)) {
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /**
+     * {@return priority main method or null if none found}
+     * @param mainClass main class
+     */
+    public static Method findMainMethod(Class<?> mainClass) throws NoSuchMethodException {
+        try {
+            Method mainMethod = mainClass.getMethod("main", String[].class);
+
+            if (mainMethod.getDeclaringClass() != mainClass) {
+                System.err.println("WARNING: static main in super class will be deprecated.");
+            }
+
+            return mainMethod;
+        } catch (NoSuchMethodException nsme) {
+            if (!PreviewFeatures.isEnabled()) {
+                throw nsme;
+            }
+
+            List<Method> mains = new ArrayList<>();
+            gatherMains(mainClass, mainClass, mains);
+
+            if (mains.isEmpty()) {
+                throw new NoSuchMethodException("No main method found");
+            }
+
+            mains.sort(MainMethodFinder::compareMethods);
```

Comment:
Perhaps skip the call to `sort` and just return the found main method if `mains.size() == 1`?